fs: shell: add cat file command

shlomow opened the pull request:
Add cat command to fs shell module.
This command prints to the standard output
the given files.

Signed-off-by: Shlomi Vaknin <shlomi.39sd@gmail.com>

de-nordic commented on subsys/fs/shell.c:
```diff
@@ -334,6 +334,66 @@ static int cmd_read(const struct shell *shell, size_t argc, char **argv)
+	create_abs_path(argv[1], path, sizeof(path));
+
+	if (argc > 2) {
+		count = strtol(argv[2], NULL, 0);
```
strtol returns long, this is implicit  truncation.

de-nordic commented on subsys/fs/shell.c:
```diff
@@ -334,6 +334,66 @@ static int cmd_read(const struct shell *shell, size_t argc, char **argv)
+	}
+
+	fs_file_t_init(&file);
+	err = fs_open(&file, path, FS_O_CREATE | FS_O_RDWR);
```
You do not need RDWR to read-only operations, you probably do not want to create file when it does not exist, which `FS_O_CREATE` will do.
The `FS_O_READ` flag suffice here. 
I know it has been done the same way in `cmd_read`, just fixed that.

de-nordic commented on subsys/fs/shell.c:
```diff
@@ -334,6 +334,66 @@ static int cmd_read(const struct shell *shell, size_t argc, char **argv)
+		int i;
+
+		read = fs_read(&file, buf, MIN(count, sizeof(buf)));
+		if (read <= 0) {
```
This does not distinguish between end of file and error conditions.
This may fail you when for example attempting to read damaged flash, here it will be visible as the file ended while actually you are not able to read past some point.

de-nordic commented on subsys/fs/shell.c:
```diff
@@ -334,6 +334,66 @@ static int cmd_read(const struct shell *shell, size_t argc, char **argv)
 	return 0;
 }
 
+static int cmd_dump(const struct shell *shell, size_t argc, char **argv)
```
Don't like the "dump" name, could it be "cat"; the "cat" name is quite common among other systems while "dump" is very often used with device operation, flash dump, memory dump, tcp dump.
Add function documentation here; your command takes parameter that needs to be figured out from code, this is not good. Some documentation for this shell function is required.
I know that currently shell lacks documentation and that would be possibly first entry, but if you at least provide doxygen for the function it will be one less function got go into to figure out the doc.

de-nordic commented on subsys/fs/shell.c:
```diff
@@ -334,6 +334,66 @@ static int cmd_read(const struct shell *shell, size_t argc, char **argv)
+
+	fs_close(&file);
+
+	return 0;
```
When `read < 0` this should return error.

de-nordic commented on subsys/fs/shell.c:
```diff
@@ -334,6 +334,66 @@ static int cmd_read(const struct shell *shell, size_t argc, char **argv)
+		return -ENOEXEC;
+	}
+
+	if (dirent.type != FS_DIR_ENTRY_FILE) {
+		shell_error(shell, "Note a file %s", path);
+		return -ENOEXEC;
+	}
```
You probably do not need this code. The `fs_open` should return error code if attempting to open non file entry. In case of directory it will be `EISDIR`.

de-nordic reviewed (changes requested):
I think that feature is OK although some changes are needed.

nvlsianpu reviewed (changes requested):
@shlomow Can you summ-up how this differ from `fs read` command?

shlomow commented:
> @shlomow Can you summ-up how this differ from `fs read` command?

Only the format of the printing, as `fs read` print also hex and ascii. As @de-nordic suggested, it makes sense to call this `cat`. I am planning to extend the implementation to support receiving multiple paths and printing them, so it will be more close to `cat`. Maybe the read command should be renamed to `hexdump`

nvlsianpu commented:
@shlomow Then I thin it make sense to extend current `fs read` command and parametrize it more.

de-nordic commented:
> 
> 
> @shlomow Then I thin it make sense to extend current `fs read` command and parametrize it more.

IMHO we should keep them separate. This way we can, in the future, add kconfig menu to enable/disable commands and shred some unused functions off the binary; someone may want to use `cat` to view logs or some text output from something, but may not need read/write.
The cat, on the other side, could altered operate on lines, in which case it could get additional options, in the future, like starting line, etc.

nvlsianpu commented:
> As @de-nordic suggested, it makes sense to call this cat. I am planning to extend the implementation to support receiving multiple paths and printing them, so it will be more close to cat.

@shlomow Ok. It can be named `cat`.

shlomow commented:
@nvlsianpu @de-nordic I updated this pr following the discussion above. Can you please review the changes?

shlomow commented:
@nvlsianpu @de-nordic ping

de-nordic commented on subsys/fs/shell.c:
```diff
@@ -334,6 +334,66 @@ static int cmd_read(const struct shell *shell, size_t argc, char **argv)
+			continue;
+		}
+
+		fs_file_t_init(&file);
```
`fs_file_t_init` is required once per `fs_file_t` object. It is supposed to allow `fs_open` to figure out whether the object is already holding info for opened file and `fs_close` to do nothing on unopened files.
Move it above the loop after variable definition block.

de-nordic commented on subsys/fs/shell.c:
```diff
@@ -334,6 +334,66 @@ static int cmd_read(const struct shell *shell, size_t argc, char **argv)
+		create_abs_path(argv[i], path, sizeof(path));
+
+		err = fs_stat(path, &dirent);
+		if (err) {
```
When checking error codes check them to be < 0.

de-nordic commented on subsys/fs/shell.c:
```diff
@@ -334,6 +334,66 @@ static int cmd_read(const struct shell *shell, size_t argc, char **argv)
+
+		while (count > 0) {
+			ssize_t read;
+			uint8_t buf[16];
```
Use some `define` macro for the buffer size and put it on top of source so that it could be configured without going through code.

de-nordic commented on subsys/fs/shell.c:
```diff
@@ -334,6 +334,66 @@ static int cmd_read(const struct shell *shell, size_t argc, char **argv)
+				break;
+			}
+
+			for (int j = 0; j < read; j++) {
+				shell_fprintf(shell, SHELL_NORMAL, "%c", buf[j]);
+			}
```
Printing each character requires invoking entire call and reinterpretation of the formatting string. Would you consider always terminating the `buf` with 0 and printing it with `"%s"` in single `shell_fprintf` call each time you are able to obtain full buffer?

shlomow replied:
@de-nordic Thanks for the review! I think that there will be a problem if the file is binary. No? I also fixed the other issues

de-nordic replied:
Ah, I have forgotten about the binary files. That is true.

de-nordic commented on subsys/fs/shell.c:
```diff
@@ -334,6 +334,66 @@ static int cmd_read(const struct shell *shell, size_t argc, char **argv)
+			continue;
+		}
+
+		count = INT_MAX;
+
+		while (count > 0) {
```
The `count` does not seem usable in any way. You do not set it as a command parameter to limit number of bytes/lines that would go out. It is used in `fs_read` call in line 373 but all it does it to limit output of file to 2^(sizeof(inf)-1) bytes.
Either make command line option or leave it out; you can check if loop should run by checking the `fs_read` still returning bytes..

de-nordic reviewed (changes requested):
Sorry for lagging.
Some changes are still needed.

de-nordic commented on subsys/fs/shell.c:
```diff
@@ -334,6 +334,62 @@ static int cmd_read(const struct shell *shell, size_t argc, char **argv)
+			uint8_t buf[BUF_CNT];
+
+			read = fs_read(&file, buf, sizeof(buf));
+			if (read == 0) {
+				break;
+			}
+
+			if (read < 0) {
+				shell_error(shell, "Failed to read from file %s (err: %d)",
+					path, read);
+				break;
+			}
```
You can break here every time `read <= 0`, and then report error when `read < 0` outside of the loop.
Currently you have two checks in each loop run that will be false most of the time.

shlomow replied:
Thanks! fixed.

de-nordic commented on subsys/fs/shell.c:
```diff
@@ -520,6 +574,9 @@ SHELL_STATIC_SUBCMD_SET_CREATE(sub_fs,
 	SHELL_CMD(pwd, NULL, "Print current working directory", cmd_pwd),
 	SHELL_CMD_ARG(read, NULL, "Read from file", cmd_read, 2, 255),
+	SHELL_CMD_ARG(cat, NULL,
+		"concatenate files and print on the standard output",
```
Upper case first letter, as all the other commands do.

shlomow replied:
Thanks! fixed.

de-nordic commented on subsys/fs/shell.c:
```diff
@@ -334,6 +334,60 @@ static int cmd_read(const struct shell *shell, size_t argc, char **argv)
+		}
+
+		err = fs_open(&file, path, FS_O_READ);
+		if (err) {
```
Sorry I missed that: `err < 0`. Generally when functions return -errno it is better to do `err < 0` to keep code more in line with processing error codes from functions that return values `>= 0` on success.

shlomow replied:
I am sorry. I should fix that when I fixed the same for calling `stat`

shlomow committed:
fs: shell: add cat file command

Add cat command to fs shell module.
This command prints to the standard output
the given files.

Signed-off-by: Shlomi Vaknin <shlomi.39sd@gmail.com>

de-nordic reviewed (approved):
Looks OK.

shlomow commented:
@nvlsianpu Can you please review this PR? :)

nvlsianpu commented:
@shlomow approved
